doc: Add release notes for 3.1.0 (#23631)
nrfconnect/sdk-nrf: doc/nrf/releases_and_maturity/releases/release-notes-3.1.0.rst

Thread samples
--------------

* Added the new :ref:`architecture option <openthread_stack_architecture>` to use the OpenThread stack directly to communicate with the IEEE 802.15.4 radio driver in the following samples:

  * :ref:`ot_coprocessor_sample`
  * :ref:`coap_server_sample`
  * :ref:`ot_cli_sample`

Wi-Fi samples
-------------

* Added:

  * The :ref:`wifi_provisioning_internal_sample` sample that demonstrates the core Wi-Fi provisioning functionality independently of the transport layer (Bluetooth LE or Wi-Fi SoftAP).
  * Runtime support for Bluetooth LE based Wi-Fi provisioning in enterprise mode.

* Removed system mode and the ``wpa_supplicant`` dependency for monitor mode optimization.

* :ref:`wifi_radiotest_samples`:

  * Updated :ref:`wifi_radio_test` and :ref:`wifi_radio_test_sd` samples to clarify platform support for single-domain and multi-domain radio tests.

* :ref:`wifi_shutdown_sample`:

  * Updated the sample to include both One-shot and Continuous modes of operations.

Other samples
-------------

* Added the :ref:`mcuboot_minimal_configuration` sample that demonstrates the minimal and recommended settings for MCUboot on the nRF54L15 DK.

Drivers
=======

> This section provides detailed lists of changes by :ref:`driver <drivers>`.

Reviewer: Place this under Drivers.

* Added the :ref:`mspi_sqspi` that allows for communication with devices that use MSPI bus-based Zephyr drivers.

Flash drivers
-------------

* Removed the SUIT flash IPUC driver.

Libraries
=========

This section provides detailed lists of changes by :ref:`library <libraries>`.

Bluetooth libraries and services
--------------------------------

* Updated the Wi-Fi Provisioning Service library by splitting it into the :ref:`lib_wifi_prov_ble` library and :ref:`lib_wifi_prov_core` library .

>   The :ref:`lib_wifi_prov_core` library is now a standalone library that can be used by applications to implement the Wi-Fi provisioning functionality independently of the Bluetooth stack.

Reviewer: Shouldn't this be under Libraries for networking?

  * Added support to provision KMU on nRF54L115, nRF54L10, and nRF54L05 devices.

* Updated the following security features:

  * HW downgrade prevention.
  * RAM memory cleanup controlled by the :kconfig:option:`CONFIG_MCUBOOT_CLEANUP_RAM` Kconfig option.
  * ECIES-X25519 encryption improvements.

* Fixed several stability and functionality improvements.

* Removed:

  * SUIT support.
  * suit-generator.
  * suit-processor.

Developing with nRF54L Series
=============================

* Added:

  * The :ref:`ug_nrf54l_dfu_config` documentation page, describing how to configure Device Firmware Update (DFU) and secure boot settings using MCUboot and NSIB.
  * The :ref:`ug_nrf54l_ecies_x25519` page on enabling the AES encryption with ECIES-X25519, which is used for secure encryption key exchange.

* Updated by increasing the default value of the :kconfig:option:`CONFIG_MPSL_HFCLK_LATENCY` Kconfig option to support slower crystals.
  See the Kconfig description for a detailed description on how to select the correct value for a given application.

Developing with nRF54H Series
=============================

> * Removed SUIT support.

Reviewer 2: Move the documentation updates here. They are so big and significant that they should be highlighted here.

  This option reduces the NVM usage of Channel Sounding when disabled by removing the ``LE CS Test`` and ``LE CS Test End`` HCI commands.

Bluetooth Mesh
--------------

* Updated:

  * The default value of the :kconfig:option:`BT_MESH_DFU_FWID_MAXLEN` Kconfig option to be compliant with the BLE Mesh specification.
  * The default value of the :kconfig:option:`BT_MESH_ADV_STACK_SIZE` Kconfig option for the nRF54L-based board targets.
  * The default value of the :kconfig:option:`MBEDTLS_HEAP_SIZE` Kconfig option if the DFU Distributor role is enabled (through the :kconfig:option:`BT_MESH_DFD_SRV` Kconfig option) to accommodate the minimum number of keys needed for the role.
  * The :ref:`bt_mesh_le_pair_resp_readme` vendor model, used in the :ref:`ble_mesh_dfu_distributor` sample, to clear Bluetooth bonding information once a node-reset message is received.

* Fixed:

  * A bug in automated metadata generation using the :kconfig:option:`SB_CONFIG_DFU_ZIP_BLUETOOTH_MESH_METADATA` Kconfig option to correctly extract vendor model metadata.
  * A buffer leakage where allocated advertisements will not be cleared after the device was suspended.

>   * A bug where the BLE scanner will be stopped when a LPN is disabled.

Reviewer 3: ```suggestion
  * A bug where the BLE scanner will be stopped when an LPN is disabled.
```

    This is only for the case when the :kconfig:option:`CONFIG_BT_MESH_LPN_ESTABLISHMENT` Kconfig option is not enabled (as this option utilizes a reduced scan cycle).
  * A bug in the :ref:`zephyr:bluetooth_mesh_blob_srv` causing the LPN queue to overflow.
    Now, if friendship is established, the maximum possible chunk size and requested number of chunks is adapted according to the current LPN queue size.

DFU samples
-----------

* Added a new sample category :ref:`dfu_samples` and a new :ref:`dfu_target_sample` sample that demonstrates the use of the Device Firmware Update (DFU) target functionality in the |NCS|.

Enhanced ShockBurst (ESB)
-------------------------

* Improved protocol disable handling in the :c:func:`esb_disable` and :c:func:`esb_stop_rx` functions.

Matter
------

* Added:

>   * FastTrack Recertification and Portfolio Certification programs.

Reviewer 2: Add link to the documentation page.

    This limits the number of GATT notifications with HID reports in the Bluetooth stack.
  * The :ref:`nrf_desktop_ble_adv_ctrl` module that is responsible for controlling the :ref:`caf_ble_adv`.
    The module suspends the |ble_adv| when the active USB device is connected (USB state is set to :c:enum:`USB_STATE_ACTIVE`).
    The module resumes the |ble_adv| when the USB is disconnected (USB state is set to :c:enum:`USB_STATE_DISCONNECTED`) and the |ble_adv| was earlier suspended.
    This improves the USB High-Speed performance.
    To enable the module, set the :ref:`CONFIG_DESKTOP_BLE_ADV_CTRL_ENABLE <config_desktop_app_options>` Kconfig option to ``y``.
    To enable the module to suspend and resume the |ble_adv| when the USB state changes, set the :ref:`CONFIG_DESKTOP_BLE_ADV_CTRL_SUSPEND_ON_USB <config_desktop_app_options>` Kconfig option to ``y``.
    These options are enabled for targets that support the USB High-Speed.
  * The following HID report providers:

    * :ref:`nrf_desktop_hid_provider_consumer_ctrl`
    * :ref:`nrf_desktop_hid_provider_keyboard`
    * :ref:`nrf_desktop_hid_provider_mouse`
    * :ref:`nrf_desktop_hid_provider_system_ctrl`

>     The HID report providers use the new :c:struct:`hid_report_provider_event` event to establish two-way callbacks used for communication with the HID state module.

Reviewer 3: ```suggestion
    The HID report providers use the new event :c:struct:`hid_report_provider_event` to establish two-way callbacks used for communication with the HID state module.
```

    The HID report providers are responsible for gathering input data and generating HID reports when requested by the HID state module.
    HID report providers can use newly added :ref:`nrf_desktop_hid_eventq`, :ref:`nrf_desktop_hid_keymap` and :ref:`nrf_desktop_keys_state` utilities.

    You can substitute existing HID report providers with your own implementation or add new HID report providers that can handle other report types.
    This change simplifies the HID state module implementation and allows code reuse.

>   * The :kconfig:option:`CONFIG_BT_ID_AUTO_SWAP_MATCHING_BONDS` Kconfig option as an imply to the :kconfig:option:`CONFIG_DESKTOP_BT_PERIPHERAL` Kconfig option.

Author: change to ref and it works.

    This helps to reduce RAM usage.
  * Link Time Optimization (:kconfig:option:`CONFIG_LTO`) to be enabled in MCUboot configurations of the nRF52840 DK (``mcuboot_smp``, ``mcuboot_qspi``).
    LTO no longer causes boot failures and it reduces the memory footprint.
  * The :ref:`nrf_desktop_hids` to use shared callbacks for multiple HID reports:

    * Use the :c:func:`bt_hids_inp_rep_send_userdata` function to send HID input reports while in report mode.
    * Use an extended callback with the notification event to handle subscriptions for HID input reports in report mode (:c:struct:`bt_hids_inp_rep`).
    * Use generic callbacks to handle HID feature and output reports.

    This approach simplifies the process of adding support for new HID reports.
  * The :ref:`nrf_desktop_hid_state` to:

    * Use the :ref:`nrf_desktop_hid_eventq` to temporarily queue HID events related to keypresses before a connection to the HID host is established.
    * Use the :ref:`nrf_desktop_hid_keymap` to map an application-specific key ID from :c:struct:`button_event` to a HID report ID and HID usage ID pair.
    * Use the :ref:`nrf_desktop_keys_state` to track the state of active keys.

>     * To offload some of its responsibilities to HID report providers modules: :ref:`nrf_desktop_hid_provider_consumer_ctrl`, :ref:`nrf_desktop_hid_provider_keyboard`, :ref:`nrf_desktop_hid_provider_mouse`, and :ref:`nrf_desktop_hid_provider_system_ctrl`.

Author: Since we have Updated heading the above suggestion doesn't suit here. Kept the texts as is.

>     * To offload some of its responsibilities to HID report providers modules: :ref:`nrf_desktop_hid_provider_consumer_ctrl`, :ref:`nrf_desktop_hid_provider_keyboard`, :ref:`nrf_desktop_hid_provider_mouse`, and :ref:`nrf_desktop_hid_provider_system_ctrl`.

Reviewer 3: Change suggested as per RMF. The bullet points could start like 'The module uses...

  * The configurations for nRF54L-based board targets that store the MCUboot verification key in the KMU peripheral to automatically generate the :file:`keyfile.json` file in the build directory (the :kconfig:option:`SB_CONFIG_MCUBOOT_GENERATE_DEFAULT_KMU_KEYFILE` sysbuild Kconfig option) based on the input file provided by the :kconfig:option:`SB_CONFIG_BOOT_SIGNATURE_KEY_FILE` sysbuild Kconfig option.
    This KMU provisioning step can now be performed automatically by the west runner, provided that a :file:`keyfile.json` file is present in the build directory.
    The provisioning is only performed if the ``west flash`` command is executed with the ``--erase``  or ``--recover`` flag.

nRF Machine Learning (Edge Impulse)
-----------------------------------

* Added:

  * Power-optimized configuration for the :zephyr:board:`nrf54h20dk` board target.
  * A single-core release configuration for the :zephyr:board:`nrf54h20dk` board target.

Serial LTE modem
----------------

* Added:

  * The ``AT#XAPOLL`` command to asynchronously poll sockets for data.

>   * The send flags for ``#XSEND``, ``#XSENDTO``, ``#XTCPSEND`` and ``#XUDPSEND`` commands.

Reviewer 4: ```suggestion
  * The send flags for ``#XSEND``, ``#XSENDTO``, ``#XTCPSEND``, and ``#XUDPSEND`` commands.
```

  * The send flag value ``512`` for waiting for acknowledgment of the sent data.

* Updated:

  * The ``AT#XPPP`` command to support the CID parameter to specify the PDN connection used for PPP.
  * The ``#XPPP`` notification to include the CID of the PDN connection used for PPP.
  * The initialization of the application to ignore a failure in nRF Cloud module initialization.
    This occurs sometimes especially during development.

>   * The initialization of the application to send "INIT ERROR" over to UART and show clear error log to indicate that the application is not operational in case of failing initialization.

Reviewer 4: ```suggestion
  * The initialization of the application to send "INIT ERROR" over UART and show clear error log to indicate that the application is not operational in case of failing initialization.
```

>   * The PPP downlink data to trigger the indicate pin when SLM is in idle.

Reviewer 4: ```suggestion
  * The PPP downlink data to trigger the indicate pin when SLM is idle.
```

    * Support for RX only mode to the ``link funmode`` command.
    * Support for ``AT%CMNG`` multi-line commands.

* :ref:`nrf_cloud_multi_service` sample:

  * Added support for native simulator platform and updated the documentation accordingly.

* :ref:`nrf_provisioning_sample` sample:

  * Updated:

    * The sample to use Zephyr's :ref:`zephyr:conn_mgr_docs` feature.
    * The sample by enabling the :ref:`lib_at_shell` library to allow the nRF Cloud Utils to interface with the device.

* :ref:`nrf_cloud_rest_device_message` sample:

  * Updated the sample to use Zephyr's :ref:`zephyr:conn_mgr_docs` feature.
  * Removed Provisioning service and JITP.

* :ref:`nrf_cloud_rest_cell_location` sample:

  * Updated the sample to use Zephyr's :ref:`zephyr:conn_mgr_docs` feature.
  * Removed JITP.

* :ref:`nrf_cloud_rest_fota` sample:

  * Updated the sample to use Zephyr's :ref:`zephyr:conn_mgr_docs` feature.
  * Fixed SMP FOTA for the nRF9160 DK.
  * Removed JITP.

Cryptography samples
--------------------

* :ref:`crypto_aes_gcm` sample:

  * Added a note stating that CRACEN only supports a 96-bit IV for AES GCM.

DFU samples
-----------

> * Added a new sample category :ref:`dfu_samples` and a new :ref:`dfu_target_sample` sample that demonstrates the use of the Device Firmware Update (DFU) target functionality in the |NCS|.

Reviewer 4: ```suggestion
* Added a new sample category :ref:`dfu_samples` and a new :ref:`dfu_target_sample` sample that demonstrates the use of the DFU target functionality in the |NCS|.
```

  * :ref:`ot_coprocessor_sample`
  * :ref:`coap_server_sample`
  * :ref:`ot_cli_sample`

Wi-Fi samples
-------------

* The :ref:`wifi_provisioning_internal_sample` sample that demonstrates the core Wi-Fi provisioning functionality independently of the transport layer (Bluetooth LE or Wi-Fi SoftAP).
* Added runtime support for Bluetooth LE based Wi-Fi provisioning in enterprise mode.
* Removed system mode and the ``wpa_supplicant`` dependency for monitor mode optimization.

* :ref:`wifi_radiotest_samples`:

  * Updated :ref:`wifi_radio_test` and :ref:`wifi_radio_test_sd` samples to clarify platform support for single-domain and multi-domain radio tests.

* :ref:`wifi_shutdown_sample`:

  * Updated the sample to include both One-shot and Continuous modes of operations.

Other samples
-------------

* Added the :ref:`mcuboot_minimal_configuration` sample that demonstrates the minimal and recommended settings for MCUboot on the nRF54L15 DK.

Drivers
=======

This section provides detailed lists of changes by :ref:`driver <drivers>`.

> * Added the :ref:`mspi_sqspi` that allows for communication with devices that use MSPI bus-based Zephyr drivers.

Author: This is already under drivers section. Am I missing something.

> * Added the :ref:`mspi_sqspi` that allows for communication with devices that use MSPI bus-based Zephyr drivers.

Reviewer: Seems to be under a subheading ADC drivers.

  This could happen, for instance, when using Bluetooth with the :kconfig:option:`CONFIG_BT_UNINIT_MPSL_ON_DISABLE` Kconfig option enabled.
  The low-frequency clock had poor accuracy in this case.

Libraries for networking
------------------------

* Added the :ref:`wifi_prov_tools` library that generates Protocol Buffers (protobuf) configuration messages for Wi-Fi provisioning, supporting both EAP-TLS (Enterprise) and Personal (WPA2-PSK/WPA3-PSK) modes.
* Updated the Wi-Fi Provisioning Service library by splitting it into the :ref:`lib_wifi_prov_ble` library and :ref:`lib_wifi_prov_core` library .
  The :ref:`lib_wifi_prov_core` library is now a standalone library that can be used by applications to implement the Wi-Fi provisioning functionality independently of the Bluetooth stack.

* :ref:`lib_nrf_cloud` library:

  * Updated:

    * To return negative :file:`errno.h` errors instead of positive ZCBOR errors.
    * The CoAP download authentication to no longer depend on the :ref:`CoAP Client library <zephyr:coap_client_interface>`.

* :ref:`lib_nrf_provisioning` library:

>   * Added

Reviewer 4: ```suggestion
  * Added:
```

The following list summarizes both the main changes inherited from upstream MCUboot and the main changes applied to the |NCS| specific additions:

* Added:

  * Support as a first-stage immutable bootloader using hardware cryptographic acceleration on nRF54L10 and nRF54L05.
  * Support for AES-256 image encryption.
  * Support for hardware downgrade protection.
  * Support for SHA512 signature verification.
  * Linking Time Optimization (LTO) support.

* Updated the following security features:

  * HW downgrade prevention.
  * RAM memory cleanup controlled by the :kconfig:option:`CONFIG_MCUBOOT_CLEANUP_RAM` Kconfig option.
  * ECIES-X25519 encryption improvements.

* Fixed:

  * Several stability and functionality improvements.
  * An issue related to referencing the ARM Vector table of the application, which causes jumping to wrong address instead of the application reset vector for some builds when Zephyr LTO (Link Time Optimization) was enabled.

Zephyr
======

.. NOTE TO MAINTAINERS: All the Zephyr commits in the below git commands must be handled specially after each upmerge and each nRF Connect SDK release.

> The Zephyr fork in |NCS| (``sdk-zephyr``) contains all commits from the upstream Zephyr repository up to and including ``9a6f116a6aa9b70b517a420247cd8d33bbbbaaa3``, with some |NCS| specific additions.

Reviewer 5: Zephyr SHA is incorrect, `0fe59bf1e4b96122c3467295b09a034e399c5ee6` is the latest merge base:
```suggestion
The Zephyr fork in |NCS| (``sdk-zephyr``) contains all commits from the upstream Zephyr repository up to and including ``0fe59bf1e4b96122c3467295b09a034e399c5ee6``, with some |NCS| specific additions.
```

For the list of upstream Zephyr commits (not including cherry-picked commits) incorporated into nRF Connect SDK since the most recent release, run the following command from the :file:`ncs/zephyr` repository (after running ``west update``):

.. code-block:: none

>    git log --oneline 9a6f116a6a ^fdeb735017

Reviewer 5: ```suggestion
   git log --oneline 0fe59bf1e4 ^fdeb735017
```

>   * Linking Time Optimization (LTO) support.

Reviewer 4: ```suggestion
  * Linking Time Optimization support.
```

For the list of |NCS| specific commits, including commits cherry-picked from upstream, run:

.. code-block:: none

>    git log --oneline manifest-rev ^9a6f116a6a

Reviewer 5: ```suggestion
   git log --oneline manifest-rev ^0fe59bf1e4
```

> The current |NCS| main branch is based on revision ``9a6f116a6a`` of Zephyr.

Reviewer 5: ```suggestion
The current |NCS| main branch is based on revision ``0fe59bf1e4`` of Zephyr.
```

>   * An issue related to referencing the ARM Vector table of the application, which causes jumping to wrong address instead of the application reset vector for some builds when Zephyr LTO (Link Time Optimization) was enabled.

Reviewer 4: ```suggestion
  * An issue related to referencing the ARM Vector table of the application, which causes jumping to wrong address instead of the application reset vector for some builds when Zephyr LTO was enabled.
```

The Zephyr fork in |NCS| (``sdk-zephyr``) contains all commits from the upstream Zephyr repository up to and including ``0fe59bf1e4b96122c3467295b09a034e399c5ee6``, with some |NCS| specific additions.

> For the list of upstream Zephyr commits (not including cherry-picked commits) incorporated into nRF Connect SDK since the most recent release, run the following command from the :file:`ncs/zephyr` repository (after running ``west update``):

Reviewer 4: ```suggestion
For the list of upstream Zephyr commits (not including cherry-picked commits) incorporated into |NCS| since the most recent release, run the following command from the :file:`ncs/zephyr` repository (after running ``west update``):
```

  * The :ref:`data_storage` page, which covers storage alternatives for general data, including NVMC, NVS, file systems, Settings, and PSA Protected Storage, with feature comparisons and configuration examples.
  * The :ref:`key_storage` page, which covers storage alternatives for cryptographic keys, including PSA Crypto API, Hardware Unique Keys (HUK), modem certificate storage, and other security-focused storage mechanisms.
  * The :ref:`migration_bicr_nrf54h` page that describes how to migrate the nRF54H20 SoC BICR configuration from DTS to JSON.
  * The :ref:`ug_nrf54h20_pm_optimization` page that describes the power management optimizations for the nRF54H20 SoC.
  * The :ref:`migration_3.1_54h_suit_ironside` page that describes how to migration from SUIT to IronSide SE for the nRF54H20 SoC.
  * The :ref:`ug_nrf54h20_ppr` page that describes how to work with the PPR core of the nRF54H20 SoC.
  * The :ref:`ug_nrf54h20_mcuboot_dfu` and :ref:`ug_nrf54h_ecies_x25519` pages that describe how to configure DFU and MCUboot on the nRF54H20 SoC.

>   * The :ref:`ug_nrf54h20_ironside` page for the nRF54H20 SoC.

Reviewer 2: I feel this should go under Developing with nRF54H20

* Updated:

  * The :ref:`ug_nrf54h20_architecture_pm` page to add a mapping of application states to software power states on the nRF54H20 SoC.

>   * The :ref:`nRF54H20 <ug_nrf54h>` documentation with the |NCS| v3.1.0 changes.

Reviewer 2: I feel this should go under Developing with nRF54H20

* Removed:

  * The Wi-Fi credentials library page and moved to the upstream :ref:`Zephyr repository <zephyr:lib_wifi_credentials>`.
  * The Getting started with nRF7002 DK and Getting started with other DKs pages from the :ref:`gsg_guides` section.
    These pages were no longer relevant as the `Quick Start app`_ now also supports the nRF7002 DK.

>   * The documentation related to SUIT.

Reviewer 2: I feel this should go under Developing with nRF54H20
```suggestion
  * The nRF54H20-specific documentation related to SUIT.
```

=============

* Added:

  * Bias-pull-up for Thingy:91 X nRF9151 UART RX pins.
  * Alternative partition tables for Thingy:91 X.

Bootloaders and DFU
===================

* Added:

  * MCUboot support as first-stage immutable bootloader using hardware cryptographic acceleration on nRF54L10 and nRF54L05.
  * MCUboot support for hardware downgrade protection the nRF54L Series devices.
  * MCUboot support for SHA512 signature verification the nRF54L Series devices.
  * Linking Time Optimization (LTO) support for MCUboot the nRF54L Series devices.
  * Experimental MCUboot support for AES-256 image encryption for the nRF54L Series devices.
  * Experimental MCUboot support as first-stage immutable bootloader for the nRF54H20:

    * Direct XIP mode with merged images for application and radio cores.
    * Swap mode with separated images for application and radio cores.

  * MCUboot sample optimized for NVM footprint and implementing recommended settings when working with the nRF54L15 DK.
    The :ref:`mcuboot_minimal_configuration` sample utilizes hardware cryptography for ED25519 signature verification and KMU for secure key storage, achieving an NVM footprint of ~16.5 KB.

>   * Added support to provision KMU on nRF54L115, nRF54L10, and nRF54L05 devices.

Reviewer: ```suggestion
  * Support to provision KMU on nRF54L115, nRF54L10, and nRF54L05 devices.
```

  See the Kconfig description for a detailed description on how to select the correct value for a given application.

Developing with nRF54H Series
=============================

* Removed SUIT support.

Developing with Thingy:91 X
===========================

* Added the :ref:`thingy91x_recover_to_factory_firmware` documentation page.

Developing with Front-End Modules
=================================

* Added:

  * The temperature compensation feature for the nRF2220 Front-End Module.
  * Support for the nRF21540 Front-End Module in GPIO/SPI mode for nRF54L Series devices.
  * Support for the Simple GPIO Front-End Module for nRF54L Series devices.

Developing with coprocessors
============================

* Added the :ref:`ug_hpf_softperipherals_comparison` documentation page, describing potential use cases and differences between the two solutions.

Security
========

* Added the new section about :ref:`ug_crypto_index`.
  The new section includes pages about :ref:`ug_crypto_architecture` (new page), :ref:`crypto_drivers` and :ref:`psa_crypto_support` (both moved from the :ref:`nrf_security` library documentation).

* Updated:

>   * Updated Mbed TLS to version 3.6.4.

Reviewer: ```suggestion
  * Mbed TLS to version 3.6.4.
```

    Now, the page only lists features and algorithms that are supported by at least one driver.
  * The Oberon PSA Crypto to version 1.5.1 that introduces support for the following new features with the Oberon PSA driver:

    * Experimental support for post-quantum cryptography schemes ML-KEM (FIPS 203), ML-DSA (FIPS 204) and NIST SP 800-208.
    * Experimental support for XChaCha20-Poly1305 AEAD algorithm.
    * Experimental support for SHAKE128 with 256-bit output length.

   The :ref:`ug_crypto_supported_features` page has been updated accordingly.

* Removed the ``CONFIG_PSA_WANT_ALG_WPA3_SAE_PT`` Kconfig option and replaced it with :kconfig:option:`CONFIG_PSA_WANT_ALG_WPA3_SAE_H2E`.

Protocols
=========

This section provides detailed lists of changes by :ref:`protocol <protocols>`.
See `Samples`_ for lists of changes for the protocol-related samples.

Bluetooth LE
------------

* Added the :kconfig:option:`CONFIG_BT_CTLR_CHANNEL_SOUNDING_TEST` Kconfig option.
  This option reduces the NVM usage of Channel Sounding when disabled by removing the ``LE CS Test`` and ``LE CS Test End`` HCI commands.

Bluetooth Mesh
--------------

* Updated:

>   * The default value of the :kconfig:option:`BT_MESH_DFU_FWID_MAXLEN` Kconfig option to be compliant with the BLE Mesh specification.

Reviewer 4: ```suggestion
  * The default value of the :kconfig:option:`BT_MESH_DFU_FWID_MAXLEN` Kconfig option to be compliant with the Bluetooth LE Mesh specification.
```

  * The default value of the :kconfig:option:`BT_MESH_ADV_STACK_SIZE` Kconfig option for the nRF54L-based board targets.
  * The default value of the :kconfig:option:`MBEDTLS_HEAP_SIZE` Kconfig option if the DFU Distributor role is enabled (through the :kconfig:option:`BT_MESH_DFD_SRV` Kconfig option) to accommodate the minimum number of keys needed for the role.
  * The :ref:`bt_mesh_le_pair_resp_readme` vendor model, used in the :ref:`ble_mesh_dfu_distributor` sample, to clear Bluetooth bonding information once a node-reset message is received.

* Fixed:

  * A bug in automated metadata generation using the :kconfig:option:`SB_CONFIG_DFU_ZIP_BLUETOOTH_MESH_METADATA` Kconfig option to correctly extract vendor model metadata.
  * A buffer leakage where allocated advertisements will not be cleared after the device was suspended.

>   * A bug where the BLE scanner will be stopped when a LPN is disabled.

Reviewer 4: ```suggestion
  * A bug where the Bluetooth LE scanner will be stopped when a LPN is disabled.
```

Enhanced ShockBurst (ESB)
-------------------------

* Improved protocol disable handling in the ``esb_disable`` and ``esb_stop_rx`` functions.

Gazell
------

|no_changes_yet_note|

Matter
------

* Added:

  * FastTrack Recertification and Portfolio Certification programs.
  * Matter-over-Thread apps can now use the OpenThread API directly, instead of using intermediate Zephyr L2 layer.
    This change significantly reduces memory usage in Matter applications.
    On the :zephyr:board:`nrf54l15dk`, it saves approximately 15 kB of RAM and 40 kB of flash.
    To learn more about the new architecture option, see the :ref:`ug_matter_networking_selection` user guide.
  * The :ref:`ug_matter_networking_selection` section on the :ref:`ug_matter_device_advanced_kconfigs` page.
    The section describes how to select the networking layer for Matter applications.

* Updated:

  * The ``west zap-generate`` command to remove previously generated ZAP files before generating new files.
    To skip removing the files, use the ``--keep-previous`` argument.
  * The :ref:`ug_matter_creating_custom_cluster` user guide by adding information about implementing custom commands.

> 

Author: Updated.

  See the :ref:`openthread_stack_architecture` user guide for more information.
  The new architecture option reduces the memory footprint of the OpenThread stack by around 4% and the RAM usage by around 12% in the :ref:`ot_cli_sample` sample.

Wi-Fi®
------

* Added support for EAP-PEAP and EAP-TTLS authentication methods to enterprise security in the Wi-Fi management API.

Applications
============

|no_changes_yet_note|

Connectivity bridge
-------------------

* Fixed to resume Bluetooth connectable advertising after a disconnect.

IPC radio firmware
------------------

|no_changes_yet_note|

Matter bridge
-------------

* Implemented the missing identify cluster for the endpoint 1.
  This resolves the :ref:`known issue <known_issues>` KRKNWK-20019.

nRF5340 Audio
-------------

* Added:

  * Experimental support for Audio on the nRF5340 DK, with LED state indications and button controls.
  * Experimental Support for stereo in :ref:`broadcast sink app<nrf53_audio_broadcast_sink_app>`.
    The broadcast sink can now receive audio from two BISes and play it on the left and right channels of the audio output, if the correct configuration options are enabled.

>     The I2S output will be stereo, but :zephyr:board:`nrf5340_audio_dk` will still only have one audio output channel, since it has a mono codec (CS47L63).

Reviewer 6: Please add to the same point:
RTT shell commands have been added to select if the HW codec shall play back I2S channel 0/left, 1/right or 0 and 1 mixed (default).

    This was due to calling the :c:func:`bt_iso_chan_send` function from a timer ISR handler and sending SDUs to the controller with invalid timestamps.

* :ref:`peripheral_hids_keyboard` and :ref:`peripheral_hids_mouse` samples:

  * Added a workaround to an issue with unexpected disconnections that resulted from improper handling of the Bluetooth Link Layer procedures by the connected Bluetooth Central device.
    This resolves the :ref:`known issue <known_issues>` NCSDK-33632.

* :ref:`nrf_auraconfig` sample:

  * Updated the buffer count (:kconfig:option:`CONFIG_BT_ISO_TX_BUF_COUNT`) to be in-line with SoftDevice Controller (SDC) defaults.
    This can be changed and optimized for specific use cases.

* :ref:`direct_test_mode` sample:

  * Fixed a bug in the workaround for errata 216 on nRF54H20 devices.
    The device asserted when a packet was received during reception tests and too few packets where transmitted during transmission tests.

* :ref:`direction_finding_peripheral` sample:

  * Added support for the ``nrf54l15dk/nrf54l15/cpuapp``, ``nrf54l15dk/nrf54l10/cpuapp``, and ``nrf54l15dk/nrf54l05/cpuapp`` board targets.

>   * Direction Finding TX AoD (atnenna switching) is disabled by default in the sample.

Reviewer: ```suggestion
  * Updated by disabling Direction Finding TX AoD (antenna switching) by default in the sample.
```

  * Updated the sample to use Zephyr's :ref:`zephyr:conn_mgr_docs` feature.
  * Removed JITP.

* :ref:`nrf_cloud_rest_fota` sample:

  * Updated the sample to use Zephyr's :ref:`zephyr:conn_mgr_docs` feature.
  * Fixed SMP FOTA for the nRF9160 DK.
  * Removed JITP.

Cryptography samples
--------------------

* :ref:`crypto_aes_gcm` sample:

  * Added a note stating that CRACEN only supports a 96-bit IV for AES GCM.

Debug samples
-------------

|no_changes_yet_note|

DECT NR+ samples
----------------

|no_changes_yet_note|

Edge Impulse samples
--------------------

|no_changes_yet_note|

Enhanced ShockBurst samples
---------------------------

|no_changes_yet_note|

Gazell samples
--------------

|no_changes_yet_note|

Keys samples
------------

|no_changes_yet_note|

Matter samples
--------------

* Added:

  * Support for the NFC onboarding for the ``nrf54l15dk/nrf54l15/cpuapp/ns`` board target.
  * Disabled usage of Zephyr L2 networking layer in favor of using the OpenThread API directly in the Matter over Thread applications.

* Updated:

  * The Bluetooth Low Energy variant of the Soft Device Controller (SDC) to use the Peripheral-only role in all Matter samples.

>   * API of the ``ncs_configure_data_model`` cmake method that does not use ``ZAP_FILE`` argument anymore, but creates path to ZAP file based on :ref:`CONFIG_NCS_SAMPLE_MATTER_ZAP_FILE_PATH <CONFIG_NCS_SAMPLE_MATTER_ZAP_FILE_PATH>` Kconfig option.

Reviewer: ```suggestion
  * The API of the ``ncs_configure_data_model`` CMake method that does not use ``ZAP_FILE`` argument anymore, but creates path to ZAP file based on :ref:`CONFIG_NCS_SAMPLE_MATTER_ZAP_FILE_PATH <CONFIG_NCS_SAMPLE_MATTER_ZAP_FILE_PATH>` Kconfig option.
```

      If the application has configured a DNS server address in Zephyr's native networking stack, using the :kconfig:option:`CONFIG_DNS_SERVER1` Kconfig option, the same server is set as the fallback address for DNS queries offloaded to the nRF91 Series modem.
      Otherwise, the :kconfig:option:`CONFIG_LTE_LC_DNS_FALLBACK_ADDRESS` Kconfig option controls the fallback DNS server address that is set to Cloudflare's DNS server 1.1.1.1 by default.
      The device might or might not receive a DNS address by the network during PDN connection.
      Even within the same network, the PDN connection establishment method (PCO vs ePCO) might change when the device operates in NB-IoT or LTE Cat-M1, resulting in missing DNS addresses when one method is used, but not the other.
      Having a fallback DNS address ensures that the device always has a DNS to fallback to.

  * Updated modem events subscription to persist between functional mode changes.

  * Removed:

    * The deprecated functions ``lte_lc_reduced_mobility_get()``, ``lte_lc_reduced_mobility_set()``, and ``lte_lc_factory_reset()``.
    * The deprecated macro ``LTE_LC_ON_CFUN()``.

* :ref:`lib_modem_slm` library:

>   * Added:

Reviewer: ```suggestion
  * Added the following Kconfig options:
```

>     * The :kconfig:option:`CONFIG_MODEM_SLM_UART_RX_BUF_COUNT` Kconfig option for configuring RX buffer count.

Reviewer: Now, I think you can remove "Kconfig option" from all items.

* :ref:`modem_info_readme` library:

  * Added:

    * The :c:func:`modem_info_get_rsrq` function for requesting the RSRQ.
    * The :c:macro:`SNR_IDX_TO_DB` macro for converting the SNR index to dB.

Multiprotocol Service Layer libraries
-------------------------------------

* Added an implementation of the API required by the MPSL (defined by :file:`mpsl_hwres.h`) for the nRF53 and nRF54L Series devices.

* Updated the implementation of the following interrupt service routine wrappers:

  * :c:func:`mpsl_timer0_isr_wrapper`
  * :c:func:`mpsl_rtc0_isr_wrapper`
  * :c:func:`mpsl_radio_isr_wrapper`

  Now, they do not trigger the kernel scheduler or use any kernel APIs.

  .. note::

     Invoking kernel APIs or triggering the kernel scheduler from Zero Latency Interrupts is considered undefined behavior.
     Users of MPSL timeslots should not assume that thread rescheduling will occur automatically at the end of a timeslot.

* Fixed an issue where calling the :c:func:`mpsl_lib_uninit` function would prevent calibration of the RC oscillator when MPSL was subsequently re-initialized using the :c:func:`mpsl_lib_init()` function.

>   This could happen, for instance, when using bluetooth with the :kconfig:option:`CONFIG_BT_UNINIT_MPSL_ON_DISABLE` Kconfig option enabled.

Reviewer: ```suggestion
  This could happen, for instance, when using Bluetooth with the :kconfig:option:`CONFIG_BT_UNINIT_MPSL_ON_DISABLE` Kconfig option enabled.
```

Integrations
============

This section provides detailed lists of changes by :ref:`integration <integrations>`.

Google Fast Pair integration
----------------------------

* Added the :ref:`ug_bt_fast_pair_adv_manager` page that describes how to integrate the :ref:`bt_fast_pair_adv_manager_readme` module in your application.
* Updated the :ref:`ug_bt_fast_pair` page to mention the availability of the guide for :ref:`ug_bt_fast_pair_adv_manager` that covers the associated helper module.
  Mentioned applicability of the :ref:`bt_fast_pair_adv_manager_readme` module in the :ref:`ug_bt_fast_pair_advertising` and the :ref:`ug_bt_fast_pair_use_case_locator_tag` sections.

Edge Impulse integration
------------------------

|no_changes_yet_note|

Memfault integration
--------------------

|no_changes_yet_note|

AVSystem integration
--------------------

|no_changes_yet_note|

nRF Cloud integration
---------------------

|no_changes_yet_note|

CoreMark integration
--------------------

|no_changes_yet_note|

DULT integration
----------------

|no_changes_yet_note|

MCUboot
=======

> 

Reviewer 7: Please add following statements under MCUBoot:

Added the following features as supported: 
•	nRF54L Series: 
     o	Bootloader and Device Firmware Update (DFU):
           •  	MCUboot as first stage immutable bootloader using hardware cryptographic acceleration on nRF54L10 and nRF54L05

Added the following features as experimental:
•	nRF54L Series: 
     o	Bootloader and Device Firmware Update (DFU):
           	    Added MCUboot support for AES-256 image encryption 
•	nRF54H20: 
     o	MCUboot as first stage immutable bootloader:
            Direct XIP mode with merged images for App and Rad Core
             Swap mode with separated images for App and Rad Core

Improved: 
•	nRF54L Series: 
       o	Bootloader and Device Firmware Update (DFU):
            	Added MCUboot sample optimized for NVM footprint and implementing recommended settings when working with nRF54L15. Sample smp_svr_mini_boot utilizes hardware cryptography for ED25519 signature verification and KMU for secure key storage, achieving NVM footprint of ~16.5KB
        	Added MCUboot support for hardware downgrade protection 
        	Added MCUboot support for SHA512 signature verification
        	Added Linking Time Optimization (LTO) support for MCUBoot

The Zephyr fork in |NCS| (``sdk-zephyr``) contains all commits from the upstream Zephyr repository up to and including ``9a6f116a6aa9b70b517a420247cd8d33bbbbaaa3``, with some |NCS| specific additions.

For the list of upstream Zephyr commits (not including cherry-picked commits) incorporated into nRF Connect SDK since the most recent release, run the following command from the :file:`ncs/zephyr` repository (after running ``west update``):

.. code-block:: none

   git log --oneline 9a6f116a6a ^fdeb735017

For the list of |NCS| specific commits, including commits cherry-picked from upstream, run:

.. code-block:: none

   git log --oneline manifest-rev ^9a6f116a6a

The current |NCS| main branch is based on revision ``9a6f116a6a`` of Zephyr.

.. note::
   For possible breaking changes and changes between the latest Zephyr release and the current Zephyr version, refer to the :ref:`Zephyr release notes <zephyr_release_notes>`.

Additions specific to |NCS|
---------------------------

|no_changes_yet_note|

zcbor
=====

|no_changes_yet_note|

Trusted Firmware-M
==================

> |no_changes_yet_note|

Reviewer 8: We can state the following

```suggestion
* Updated to TF-M version 2.1.2 which is a bugfix release which addresses issues in protected storage and in the secure partition manager. This release is integrated towards Mbed TLS v3.6.4
```

    Then, after the :ref:`CONFIG_DESKTOP_USB_PM_RESTRICT_REMOVE_DELAY_MS <config_desktop_app_options>` configurable delay, the module removes the power down level restriction.
    This allows you to take actions, such as restart Bluetooth LE advertising, after disconnecting the USB cable without going through reboot.
  * The configurations for nRF54L-based board targets that store the MCUboot verification key in the KMU peripheral to automatically generate the :file:`keyfile.json` file in the build directory (the :kconfig:option:`SB_CONFIG_MCUBOOT_GENERATE_DEFAULT_KMU_KEYFILE` sysbuild Kconfig option) based on the input file provided by the :kconfig:option:`SB_CONFIG_BOOT_SIGNATURE_KEY_FILE` sysbuild Kconfig option.
    This KMU provisioning step can now be performed automatically by the west runner, provided that a :file:`keyfile.json` file is present in the build directory.
    The provisioning is only performed if the ``west flash`` command is executed with the ``--erase``  or ``--recover`` flag.

nRF Machine Learning (Edge Impulse)
-----------------------------------

* Added power-optimized configuration for the :zephyr:board:`nrf54h20dk` board target.

> 

Reviewer 9: Added a single-core release configuration for the :zephyr:board:`nrf54h20dk` board target.

  * :ref:`https_client`
  * :ref:`mqtt_sample`
  * :ref:`udp_sample`
  * :ref:`http_server`

* :ref:`download_sample` sample:

  * Added the :ref:`CONFIG_SAMPLE_PROVISION_CERT <CONFIG_SAMPLE_PROVISION_CERT>` Kconfig option to provision the root CA certificate to the modem.
    The certificate is provisioned only if the :ref:`CONFIG_SAMPLE_SECURE_SOCKET <CONFIG_SAMPLE_SECURE_SOCKET>` Kconfig option is set to ``y``.
  * Fixed an issue where the network interface was not re-initialized after a fault.

NFC samples
-----------

* Added experimental ``llvm`` toolchain support for the ``nrf54l15dk/nrf54l15/cpuapp`` board target to the following samples:

  * :ref:`writable_ndef_msg`
  * :ref:`nfc_shell`

* :ref:`record_text` sample:

  * Added support for the ``nrf54l15dk/nrf54l15/cpuapp/ns`` board target.

nRF5340 samples
---------------

|no_changes_yet_note|

Peripheral samples
------------------

* :ref:`radio_test` sample:

  * Added experimental ``llvm`` toolchain support for the ``nrf54l15dk/nrf54l15/cpuapp`` board target.

* :ref:`802154_phy_test` sample:

  * Added print of sent packets and received Acks after ``ltx`` command.

PMIC samples
------------

> |no_changes_yet_note|

Reviewer 10: please add:
```rst
* :ref:`npm1300_fuel_gauge` sample:
  * Changed current sign when calling the ``nrf_fuel_gauge`` lib functions.
    In the Zephyr sensor API the convention for the gauge current is negative=discharging, while nrf_fuel_gauge lib uses the opposite
```

The Zephyr fork in |NCS| (``sdk-zephyr``) contains all commits from the upstream Zephyr repository up to and including ``9a6f116a6aa9b70b517a420247cd8d33bbbbaaa3``, with some |NCS| specific additions.

For the list of upstream Zephyr commits (not including cherry-picked commits) incorporated into nRF Connect SDK since the most recent release, run the following command from the :file:`ncs/zephyr` repository (after running ``west update``):

.. code-block:: none

   git log --oneline 9a6f116a6a ^fdeb735017

For the list of |NCS| specific commits, including commits cherry-picked from upstream, run:

.. code-block:: none

   git log --oneline manifest-rev ^9a6f116a6a

The current |NCS| main branch is based on revision ``9a6f116a6a`` of Zephyr.

.. note::
   For possible breaking changes and changes between the latest Zephyr release and the current Zephyr version, refer to the :ref:`Zephyr release notes <zephyr_release_notes>`.

Additions specific to |NCS|
---------------------------

|no_changes_yet_note|

zcbor
=====

|no_changes_yet_note|

Trusted Firmware-M
==================

|no_changes_yet_note|

cJSON
=====

|no_changes_yet_note|

Documentation
=============

> * Added:
> 
>   * The :ref:`asset_tracker_template_redirect` page, which provides the information about the `Asset Tracker Template Add-on <Asset Tracker Template_>`_.
>   * The :ref:`log_rpc` library documentation page.
>   * The :ref:`mcuboot_serial_recovery` documentation page, based on the official Zephyr documentation, which discusses the implementation and usage of the serial recovery.
>   * The :ref:`data_storage` page, which covers storage alternatives for general data, including NVMC, NVS, file systems, Settings, and PSA Protected Storage, with feature comparisons and configuration examples.
>   * The :ref:`key_storage` page, which covers storage alternatives for cryptographic keys, including PSA Crypto API, Hardware Unique Keys (HUK), modem certificate storage, and other security-focused storage mechanisms.
>   * The :ref:`bt_fast_pair_adv_manager_readme` page that describes the new helper module for the :ref:`bt_fast_pair_readme` library.
>   * The :ref:`migration_bicr_nrf54h` page that describes how to migrate the nRF54H20 SoC BICR configuration from DTS to JSON.
> 
> * Updated:
> 
>   * The :ref:`bt_fast_pair_readme` page to mention the availability of the :ref:`bt_fast_pair_adv_manager_readme` helper module.
>   * Moved the Wi-Fi credentials library page to the upstream :ref:`Zephyr repository <zephyr:lib_wifi_credentials>`.

Reviewer 11: ```suggestion
* Added:

  * The :ref:`asset_tracker_template_redirect` page, which provides the information about the `Asset Tracker Template Add-on <Asset Tracker Template_>`_.
  * The :ref:`log_rpc` library documentation page.
  * The :ref:`mcuboot_serial_recovery` documentation page, based on the official Zephyr documentation, which discusses the implementation and usage of the serial recovery.
  * The :ref:`data_storage` page, which covers storage alternatives for general data, including NVMC, NVS, file systems, Settings, and PSA Protected Storage, with feature comparisons and configuration examples.
  * The :ref:`key_storage` page, which covers storage alternatives for cryptographic keys, including PSA Crypto API, Hardware Unique Keys (HUK), modem certificate storage, and other security-focused storage mechanisms.
  * The :ref:`bt_fast_pair_adv_manager_readme` page that describes the new helper module for the :ref:`bt_fast_pair_readme` library.
  * The :ref:`migration_bicr_nrf54h` page that describes how to migrate the nRF54H20 SoC BICR configuration from DTS to JSON.
  * The :ref:`ug_nrf54h20_pm_optimization` page describing nRF54H20 power management optimization.
  * The :ref:`migration_3.1_54h_suit_ironside` page describing Migration from SUIT to IronSide SE for the nRF54H20 SoC.
  * The :ref:`ug_nrf54h20_ppr` page describing Working with the PPR core.
  * The :ref:`ug_nrf54h20_mcuboot_dfu` and :ref:`ug_nrf54h_ecies_x25519` pages describing how to configure DFU and MCUboot on the nRF54H20 SoC.

* Updated:

  * The :ref:`bt_fast_pair_readme` page to mention the availability of the :ref:`bt_fast_pair_adv_manager_readme` helper module.
  * Moved the Wi-Fi credentials library page to the upstream :ref:`Zephyr repository <zephyr:lib_wifi_credentials>`.
  * The :ref:`ug_nrf54h20_architecture_pm` page.
  * The :ref:`ug_nrf54h20_flpr` page.
```

Reviewer 11: Also this https://github.com/nrfconnect/sdk-nrf/pull/22940 and this https://github.com/nrfconnect/sdk-nrf/pull/22754 will need an entry when merged

  * Updated current sign when calling the ``nrf_fuel_gauge`` lib functions.
    In the Zephyr sensor API the convention for the gauge current is ``negative=discharging``, while the :ref:`nrfxlib:nrf_fuel_gauge` lib uses the opposite.

Protocol serialization samples
------------------------------

|no_changes_yet_note|

SDFW samples
------------

|no_changes_yet_note|

Sensor samples
--------------

|no_changes_yet_note|

SUIT samples
------------

* Removed all SUIT samples:

  * SUIT: Device firmware “A/B” update on the nRF54H20 SoC
  * SUIT: Flash companion
  * SUIT: Recovery application
  * SUIT: Device firmware update on the nRF54H20 SoC

Trusted Firmware-M (TF-M) samples
---------------------------------

* :ref:`tfm_secure_peripheral_partition` sample:

  * Added support for the ``nrf54l15dk/nrf54l15/cpuapp/ns`` board target.

Thread samples
--------------

* Added the new :ref:`architecture option <openthread_stack_architecture>` to use the OpenThread stack directly to communicate with the IEEE 802.15.4 radio driver in the following samples:

  * :ref:`ot_coprocessor_sample`
  * :ref:`coap_server_sample`
  * :ref:`ot_cli_sample`

Wi-Fi samples
-------------

> * Added runtime support for Buletooth LE based Wi-Fi provisioning in enterprise mode.

Reviewer 4: Please add under the Added section for https://github.com/nrfconnect/sdk-nrf/pull/23552 PR, (@krish2718 to confirm) :

* The :ref:`wifi_provisioning_internal_sample` sample that demonstrates the core Wi-Fi provisioning functionality independently of the transport layer (Bluetooth LE or Wi-Fi SoftAP).

* Removed system mode and the ``wpa_supplicant`` dependency for monitor mode optimization.

* :ref:`wifi_radiotest_samples`:

  * Updated :ref:`wifi_radio_test` and :ref:`wifi_radio_test_sd` samples to clarify platform support for single-domain and multi-domain radio tests.

* :ref:`wifi_shutdown_sample`:

  * Updated the sample to include both One-shot and Continuous modes of operations.

Other samples
-------------

* Added the :ref:`mcuboot_minimal_configuration` sample that demonstrates the minimal and recommended settings for MCUboot on the nRF54L15 DK.

Drivers
=======

This section provides detailed lists of changes by :ref:`driver <drivers>`.

* Added the :ref:`mspi_sqspi` that allows for communication with devices that use MSPI bus-based Zephyr drivers.

Wi-Fi drivers
-------------

|no_changes_yet_note|

Flash drivers
-------------

* Removed the SUIT flash IPUC driver.

Libraries
=========

This section provides detailed lists of changes by :ref:`library <libraries>`.

Binary libraries
----------------

|no_changes_yet_note|

> Bluetooth libraries and services

Reviewer 4: Please add for https://github.com/nrfconnect/sdk-nrf/pull/23552 PR, (@krish2718 to confirm):

* Updated the Wi-Fi Provisioning Service library by splitting it into the :ref:`lib_wifi_prov_ble` library and :ref:`lib_wifi_prov_core` library .
    The :ref:`lib_wifi_prov_core` library is now a standalone library that can be used by applications to implement the Wi-Fi provisioning functionality independently of the Bluetooth stack.

Multiprotocol Service Layer libraries
-------------------------------------

* Added an implementation of the API required by the MPSL (defined by :file:`mpsl_hwres.h`) for the nRF53 and nRF54L Series devices.

* Updated the implementation of the following interrupt service routine wrappers:

  * :c:func:`mpsl_timer0_isr_wrapper`
  * :c:func:`mpsl_rtc0_isr_wrapper`
  * :c:func:`mpsl_radio_isr_wrapper`

  Now, they do not trigger the kernel scheduler or use any kernel APIs.

  .. note::

     Invoking kernel APIs or triggering the kernel scheduler from Zero Latency Interrupts is considered undefined behavior.
     Users of MPSL timeslots should not assume that thread rescheduling will occur automatically at the end of a timeslot.

* Fixed an issue where calling the :c:func:`mpsl_lib_uninit` function would prevent calibration of the RC oscillator when MPSL was subsequently re-initialized using the :c:func:`mpsl_lib_init` function.

  This could happen, for instance, when using Bluetooth with the :kconfig:option:`CONFIG_BT_UNINIT_MPSL_ON_DISABLE` Kconfig option enabled.
  The low-frequency clock had poor accuracy in this case.

> Libraries for networking

Reviewer 4: Please add for https://github.com/nrfconnect/sdk-nrf/pull/23552 PR, (@krish2718 to confirm):

* Added the :ref:`wifi_prov_tools` library that generates Protocol Buffers (protobuf) configuration messages for Wi-Fi provisioning, supporting both EAP-TLS (Enterprise) and Personal (WPA2-PSK/WPA3-PSK) modes.
